Unable to use vendoring with golang-http

<!-- Due dilligence -->

## Why do you need this?
<!--- Explain how this is affecting you, or what you're trying to achieve -->

I need to be able to use vendoring with the golang-http template because it is how I get private Go code into my functions. 

## Expected Behaviour
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->

This should work without error, when using a vendor folder and any other stack.yml flags or build-args required:

```
faas-cli build
```

## Current Behaviour
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

```
Step 29/42 : RUN ${GO} build --ldflags "-s -w" -a -installsuffix cgo -o handler .
 ---> Running in b9e5d187d314
Setting vendor mode env variables
main.go:18:2: cannot find package "github.com/openfaas/templates-sdk/go-http" in any of:
        /usr/local/go/src/github.com/openfaas/templates-sdk/go-http (from $GOROOT)
        /go/src/github.com/openfaas/templates-sdk/go-http (from $GOPATH)
The command '/bin/sh -c ${GO} build --ldflags "-s -w" -a -installsuffix cgo -o handler .' returned a non-zero code: 1
[0] < Building go1 done in 5.91s.
[0] Worker done.

Total build time: 5.91s
Errors received during build:
- [go1] received non-zero exit code from build, error: The command '/bin/sh -c ${GO} build --ldflags "-s -w" -a -installsuffix cgo -o handler .' returned a non-zero code: 1
```

## List All Possible Solutions and Workarounds
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->

1) Create a new separate template `golang-http-vendoring` that only works with vendoring
2) Stop maintaining a golang-http template at all, migrate everyone to golang-middleware
3) Use interfaces instead of structs, so that the main.go module doesn't have to reference the SDK, which seems to be the source of the problem - full example, and it's a breaking change: https://github.com/alexellis/golang-http-template
4) Consider a new direction that looks like AWS Lambda's Go handler - https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html
5) Use Go's plugin capability, to build the function and entry point separately, and link them after the fact. Unfortunately this seems to require CGO and breaks portability - https://github.com/golang/go/issues/19569

For 4), the function's handler owns the main() entrypoint instead of the hidden underlying code, and calls a utility method to set up the HTTP server etc.

```golang
package main

import (
        "fmt"
        "context"
        "github.com/openfaas/golang-http/entrypoint"
)

type MyEvent struct {
        Name string `json:"name"`
}

func HandleRequest(ctx context.Context, name MyEvent) (string, error) {
        return fmt.Sprintf("Hello %s!", name.Name ), nil
}

func main() {
        entrypoint.Start(HandleRequest)
}
```

For 3), the code would end up looking like this for a handler:

(this can be compared to the second code sample to see what's different)

```golang
package function

import (
    "fmt"
    "net/http"

    gohttp "github.com/alexellis/templates-sdk/go-http"
    logrus "github.com/sirupsen/logrus"
)

// Handle a function invocation
func Handle(req gohttp.Request) (gohttp.Response, error) {
    var err error

    logrus.Info("Value of body: ", string(req.GetBody()))

    message := fmt.Sprintf("Body: %s", string(req.GetBody()))

    return &gohttp.FunctionResponse{
        Body:       []byte(message),
        StatusCode: http.StatusOK,
    }, err
}
```

## What is your preferred solution?

I like 3) and to lessen its impact to existing users, we could consider promoting golang-http into the primary community templates repository and deprecating the one we use here.

Alternatively, we could call it "go-http1.18` or `golang-http1.18` as a new template, with `golang-http1.19` and so on coming after that.

## Steps to Reproduce (for bugs)

```bash
faas-cli template store pull golang-http-template
faas-cli new --lang golang-http  --prefix alexellis2 withlogrus
```

Edit `withlogrus/handler.go`:

```golang
package function

import (
	"fmt"
	"io"
	"net/http"

	"github.com/sirupsen/logrus"
)

func Handle(w http.ResponseWriter, r *http.Request) {
	var input []byte

	if r.Body != nil {
		defer r.Body.Close()

		body, _ := io.ReadAll(r.Body)

		input = body
	}

	logrus.Debug(fmt.Sprintf("Handle; %q", input))

	w.WriteHeader(http.StatusOK)
	w.Write([]byte(fmt.Sprintf("Body: %q", string(input))))
}
```

Then edit `withlogrus.yml`, adding:

```
    build_args:
      GO111MODULE: off
      GOFLAGS: "-mod=vendor"
```

Then try a build:

```
cd golang-http-template

go get
go mod vendor

cd ..
faas-cli build -f withlogrus.yml
```

--- a/go.work
+++ b/go.work
@@ -1,6 +1,0 @@
-go 1.18
-
-use (
-	./template/golang-http
-	./template/golang-middleware
-)

--- a/template/golang-http/Dockerfile
+++ b/template/golang-http/Dockerfile
@@ -26,13 +26,16 @@ ENV CGO_ENABLED=${CGO_ENABLED}
 # Run a gofmt and exclude all vendored code.
 RUN test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*" -not -path "./function/vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
 
+RUN sh ./modules-cleanup.sh
+
 WORKDIR /go/src/handler/function
+
 RUN mkdir -p /go/src/handler/function/static
 
-RUN GOOS=${TARGETOS} GOARCH=${TARGETARCH} go test ./... -cover
+RUN GOOS=${TARGETOS} GOARCH=${TARGETARCH} GOFLAGS=${GOFLAGS} go test ./... -cover
 
 WORKDIR /go/src/handler
-RUN CGO_ENABLED=${CGO_ENABLED} GOOS=${TARGETOS} GOARCH=${TARGETARCH} \
+RUN CGO_ENABLED=${CGO_ENABLED} GOOS=${TARGETOS} GOARCH=${TARGETARCH} GOFLAGS=${GOFLAGS} \
     go build --ldflags "-s -w" -a -installsuffix cgo -o handler .
 
 FROM --platform=${TARGETPLATFORM:-linux/amd64} alpine:3.16.2 as ship

--- a/template/golang-http/function/go.mod
+++ b/template/golang-http/function/go.mod
@@ -1,3 +1,5 @@
 module handler/function
 
 go 1.18
+
+require github.com/openfaas/templates-sdk/go-http v0.0.0-20220408082716-5981c545cb03

--- a/template/golang-http/go.work
+++ b/template/golang-http/go.work
@@ -1,6 +1,0 @@
-go 1.18
-
-use (
-	.
-	./function
-)

--- a/template/golang-http/modules-cleanup.sh
+++ b/template/golang-http/modules-cleanup.sh
@@ -1,0 +1,149 @@
+#!/usr/bin/env sh
+
+set -e
+
+GO111MODULE=$(go env GO111MODULE)
+
+# move_vendor will copy the function's vendor folder,
+# if it exists.
+move_vendor() {
+    if [ ! -d ./function/vendor ]; then
+        echo "vendor not found"
+        return
+    fi
+
+    echo "moving function vendor"
+    mv -f ./function/vendor .
+}
+
+
+# cleanup_gomod will move the function's go module
+cleanup_gomod() {
+
+    # Nothing to do when modules is explicitly off
+    # the z prefix protects against any SH wonkiness
+    # see https://stackoverflow.com/a/18264223
+    if [ "z$GO111MODULE" = "zoff" ]; then
+        echo "modules disabled, skipping go.mod cleanup"
+        return;
+    fi
+
+    if [ ! -f ./function/go.mod ]; then
+        echo "module not initialized, skipping go.mod cleanup"
+        return;
+    fi
+
+    echo "cleaning up go.mod"
+
+    # Copy the user's go.mod
+    mv -f ./function/go.mod .
+    mv -f ./function/go.sum .
+
+    # Clean up the go.mod
+
+    # Cleanup any sub-module replacements.
+    # This requires modifying any replace that points to "./*",
+    # the user has will use this to reference sub-modules instead
+    # of sub-packages, which we cleanup below.
+    echo "cleanup local replace statements"
+    # 1. Replace references to the local folder with `./function`
+    sed -i 's/=> \.\//=> \.\/function\//' go.mod
+
+
+    # Remove any references to the handler/function module.
+    # It is ok to just remove it because we will replace it later.
+    #
+    # Note that these references may or may not exist. We expect the
+    # go.mod to have a replace statement _if_  developer has subpackages
+    # in their handler. In this case they will need a this replace statement
+    #
+    #    replace handler/function => ./
+    #
+    # `go mod` will then add a line that looks like
+    #
+    #    handler/function v0.0.0-00010101000000-000000000000
+    #
+    # both of these lines need to be replaced, this grep selects everything
+    # _except_ those offending lines.
+    grep -v "\shandler/function" go.mod > gomod2; mv gomod2 go.mod
+
+    # Now update the go.mod
+    #
+    # 1. use replace so that imports of handler/function use the local code
+    # 2. we need to rename the module to handler because our main.go assumes
+    #    this is the package name
+    go mod edit \
+        -replace=handler/function=./function \
+        -module handler
+
+
+
+    if [ "$DEBUG" -eq 1 ]; then
+        cat go.mod
+        echo ""
+    fi
+}
+
+
+# cleanup_vendor_modulestxt will cleanup the modules.txt file in the vendor folder
+# this file is needed when modules are enabled and it must be in sync with the
+# go.mod.  To function correctly we need to modify the references to handler/function,
+# if they exist.
+cleanup_vendor_modulestxt() {
+    if [ ! -d ./vendor ]; then
+        echo "no vendor found, skipping modules.txt cleanup"
+        return
+    fi
+
+    # Nothing to do when modules is explicitly off
+    # the z prefix protects against any SH wonkiness
+    # see https://stackoverflow.com/a/18264223
+    if [ "z$GO111MODULE" = "zoff" ]; then
+        echo "modules disabled, skipping modules.txt cleanup"
+        return;
+    fi
+
+    echo "cleanup vendor/modules.txt"
+
+    # just in case
+    touch "./vendor/modules.txt"
+
+    # when vendored, we need to do similar edits to the vendor/modules.txt
+    # as we did to the go.mod
+
+    # 1. we need to replace any possible copy of the handler code
+    rm -rf vendor/handler && \
+
+    # 2. in modules.txt, we remove existing references to the handler/function
+    #    we reconstruct these in the last step
+    grep -v "\shandler/function" ./vendor/modules.txt> modulestext; mv modulestext ./vendor/modules.txt
+
+    # 3. Handle any other local replacements.
+    # any replace that points to `./**` needs to be udpat    echo "cleanup local replace statements"
+    sed -i 's/=> \.\//=> \.\/function\//' ./vendor/modules.txt
+
+    # 4. To make the modules.txt consistent with the new go.mod,
+    #    we add the mising replace to the vendor/modules.txt
+    echo "## explicit" >> ./vendor/modules.txt
+    echo "# handler/function => ./function" >> ./vendor/modules.txt
+
+    if [ "$DEBUG" -eq 1 ]; then
+        cat ./vendor/modules.txt;
+        echo ""
+    fi
+}
+
+# has_local_replacement checks if the file contains local go module replacement
+has_local_replacement() {
+    return "$(grep -E -c '=> \./\S+' "$1")"
+}
+
+
+################
+#   main
+################
+move_vendor
+
+cleanup_gomod
+
+cleanup_vendor_modulestxt

